dfdaemon/config/config.go

```diff
@@ -114,6 +114,11 @@ type Properties struct {
 	LocalIP    string          `yaml:"localIP" json:"localIP"`
 	PeerPort   int             `yaml:"peerPort" json:"peerPort"`
 	StreamMode bool            `yaml:"streamMode" json:"streamMode"`
+
+	ProtocolConf []ProtocolConfig `yaml:"protocolConf" json:"protocolConf"`
+	// DefaultPattern is the default pattern.
+	DefaultPattern string          `yaml:"defaultPattern" json:"defaultPattern"`
+	PatternConf    []PatternConfig `yaml:"patternConf" json:"patternConf"`
```

Comment:
Instead of having a default pattern key as specified above, does it also make sense to just assuming the first one in this list is the default one?

Reply:
> Instead of having a default pattern key as specified above, does it also make sense to just assuming the first one in this list is the default one?

If we always take the first one as default, then you need to change the order of `PatternConf` to change the default, which is kind of wierd. `PatternConf` should mean "what is available". And `PatternConf` might also be empty for a minimal quick start configuration.

Reply:
Just thinking loud, the default value `p2p` as mentioned in setup functions, what happens if there is no entry for `p2p` here?

And also from an operation perspective, the way I need to pick a key from another part just makes it more error prone - typo, non-existent, etc.

And really, do we really need to provide a default one? If client doesn't spell out it explicitly, it should be reasonable to report a failure, no?

The default pattern, as far as I understand, is really a concern from client perspective.

Reply:
In our design, we provide two ways to choose pattern for one request. One is static selection, which is configuration from config file as  the default pattern; Another is dynamic way,  in which user could choose pattern by specify param in request, and by patternMatcher specific request will run in specific pattern.